server/storage/database.go

```diff
@@ -1,177 +1,165 @@
+	"fmt"
 
 	"github.com/Sirupsen/logrus"
 	"github.com/endophage/gotuf/data"
-	"github.com/go-sql-driver/mysql"
+	"github.com/jinzhu/gorm"
 )
 
-// MySQLStorage implements a versioned store using a relational database.
-// The database table must look like:
-// CREATE TABLE `tuf_files` (
-//   `id` INT AUTO_INCREMENT,
-//   `gun` VARCHAR(255) NOT NULL
-//   `role` VARCHAR(255) NOT NULL
-//   `version` INT
-//   `data` LONGBLOB
-//   PRIMARY KEY (`id`)
-//   UNIQUE INDEX (`gun`, `role`, `version`)
-// ) DEFAULT CHARSET=utf8;
-//
-// CREATE TABLE `timestamp_keys` (
-//   `gun` VARCHAR(255),
-//   `cipher` VARCHAR(30),
-//   `public` BLOB NOT NULL,
-// ) DEFAULT CHARSET=utf8;
-type MySQLStorage struct {
-	sql.DB
+// SQLStorage implements a versioned store using a relational database.
+// See server/storage/models.go
+type SQLStorage struct {
+	gorm.DB
 }
 
-// NewMySQLStorage is a convenience method to create a MySQLStorage
-func NewMySQLStorage(db *sql.DB) *MySQLStorage {
-	return &MySQLStorage{
-		DB: *db,
+// NewSQLStorage is a convenience method to create a SQLStorage
+func NewSQLStorage(dialect string, args ...interface{}) (*SQLStorage, error) {
+	gormDB, err := gorm.Open(dialect, args...)
+	if err != nil {
+		return nil, err
 	}
+	return &SQLStorage{
+		DB: gormDB,
+	}, nil
 }
 
-// UpdateCurrent updates multiple TUF records in a single transaction.
-// Always insert a new row. The unique constraint will ensure there is only ever
-func (db *MySQLStorage) UpdateCurrent(gun string, update MetaUpdate) error {
-	checkStmt := "SELECT count(*) FROM `tuf_files` WHERE `gun`=? AND `role`=? AND `version`>=?;"
-	insertStmt := "INSERT INTO `tuf_files` (`gun`, `role`, `version`, `data`) VALUES (?,?,?,?);"
+// UpdateCurrent updates a single TUF.
+func (db *SQLStorage) UpdateCurrent(gun string, update MetaUpdate) error {
+	// ensure we're not inserting an immediately old version - can't use the
+	// struct, because that only works with non-zero values, and Version
+	// can be 0.
+	exists := db.Where("gun = ? and role = ? and version >= ?",
+		gun, update.Role, update.Version).First(&TUFFile{})
 
-	// ensure we're not inserting an immediately old version
-	row := db.QueryRow(checkStmt, gun, update.Role, update.Version)
-	var exists int
-	err := row.Scan(&exists)
-	if err != nil {
-		return err
-	}
-	if exists != 0 {
+	if !exists.RecordNotFound() {
 		return &ErrOldVersion{}
 	}
 
-	// attempt to insert. Due to race conditions with the check this could fail.
-	// That's OK, we're doing first write wins. The client will be messaged it
-	// needs to rebase.
-	_, err = db.Exec(insertStmt, gun, update.Role, update.Version, update.Data)
-	if err != nil {
-		if err, ok := err.(*mysql.MySQLError); ok {
-			if err.Number == 1022 { // duplicate key error
-				return &ErrOldVersion{}
-			}
-		}
-		// need to check error type for duplicate key exception
-		// and return ErrOldVersion if duplicate
-		return err
-	}
-	return nil
+	return db.Create(&TUFFile{
```

Comment:
Might be good to capture the error and check a key error in case a race condition has caused this to be written by another process. Will allow us to raise the ErrOldVersion and handle it more consistently.